Cesium.CodeGen/Contexts/IDeclarationScope.cs

```diff
@@ -11,5 +11,7 @@ internal interface IDeclarationScope
     IReadOnlyDictionary<string, IType> Variables { get; }
     IReadOnlyDictionary<string, IType> GlobalFields { get; }
     void AddVariable(string identifier, IType variable);
+    IType DisambiguateType(IType type);
```

Comment:
Could you please explain what this method is supposed to do? Or maybe even add the xmldoc here.

Reply:
This method find real C type based on the name of type or return existing if all type information already captured. 

I believe this should be separate stage of approximately following pipeline
1. Registration of declarations
2. Type resolution (this method and maybe something else)
3. Code transformation(Lower)
4. Emit 

I also suspect that we choose Lower as method name incorrectly as I think in compiler development this means very specific stage.

Reply:
So, this is actually `Resolve`?

To me, the signature looks pretty strange: why does it take an `IType`? If I want a type, then presumably I shouldn't have one?

Reply:
Technically we have only NamedType which is half baked, but it can be anywhere in the concrete ITypes

Reply:
What if we limit this to only accept `NamedType`? Or even just to accept the name? This would force the reader to figure out what to do with the rest of the type.

Reply:
We need to resolve StructType in which NamedType are type of the field for example. Or if NamedType is base type for array. 

This function is almost type resolution phase which I cannot do now. Currently I at least two small refactoring away from talking with you about clear stages. First step is conversion of ScopedIdentifierDeclaration into IBlockDeclaration. Then do the same for FunctionDeclaration and probably some wrap up of what’s left. After that we can work on separate phases. That’s mundane but not obvious work with delayed benefits. At least after hiding stage in the function I can unblock one potential contributor and couple new work items. 

But if you have ideas what I can do better, I willing to change something or go into other direction.

Reply:
Okay. After re-reading the imlementation, I believe it is now correct and no, we cannot limit it to `NamedType`.

I suggest naming it either `FullyResolve`, or `RecursivelyResolve`, or `DeepResolve`, because that's what it really does in compiler terms: it does name resolve for the type itself and for all its members recursively.

Another naming variant, if you still insist on "resolve" being something else, then we can call it `Materialize` or `Instantiate`.

And we should add an explaining comment to this method. I suggest something like this:
```
/// Recursively resolve the passed type and all its members, replacing `NamedType` in any points with their actual instantiations in the current context.
///
/// Throws a <see cref="CompilationException"/> if it's not possible to resolve some of the types.
```

Reply:
I would vote for `ResolveType` since we cannot resolve type "not fully". There circumstances where our code can work on unresolved types, like some usages of pointers like `foo*` but that's mean that we do not resolve type yet.

Reply:
Alright, that will work.